Fix #202: Add support for composite unique constraints

vishwamartur opened the pull request:
## Summary

This PR resolves GitHub issue #202 by implementing comprehensive support for composite unique constraints across all datasources. The issue was that the API only checked individual column uniqueness but didn't properly handle composite unique constraints (multiple columns together forming a unique constraint).

## Problem Solved

Before this fix:
- The system would run separate `SELECT COUNT(*)` queries for each individual column
- For a composite constraint like `UNIQUE(user_id, server_id)`, it would check `user_id` and `server_id` separately
- This caused false positives where a user could have multiple reviews for different servers, but the individual `user_id` check would fail

After this fix:
- The system detects composite unique constraints from the database schema
- Validates the combination of columns together using proper `WHERE user_id = ? AND server_id = ?` queries
- Provides clear error messages indicating which combination of values already exists

## Changes Made

### 1. Core Types Enhancement
- Added `DataSourceCompositeUniqueConstraint` interface
- Extended `DataSourceSchema` to include `composite_unique_constraints` field

### 2. Database Schema Discovery
- **PostgreSQL**: Query `information_schema` to detect composite unique constraints
- **MySQL**: Query `INFORMATION_SCHEMA` to detect composite unique constraints  
- **SQL Server**: Query `INFORMATION_SCHEMA` to detect composite unique constraints
- **MongoDB**: Added empty array for consistency (NoSQL doesn't use SQL constraints)
- **Airtable**: Added empty array for consistency (doesn't support composite constraints)

### 3. Uniqueness Validation Enhancement
- Enhanced `uniqueCheck` method in all SQL datasources
- Removed early returns that prevented composite constraint checking
- Added logic to build proper WHERE clauses for multi-column constraints
- Support for updates by excluding current record ID from constraint checks

### 4. Comprehensive Test Suite
- Unit tests for PostgreSQL, MySQL, and SQL Server composite constraint validation
- Integration tests specifically testing the GitHub issue #202 scenario
- Tests for reviews table with `user_id` and `server_id` composite constraint
- Verification of constraint violation detection, updates, and edge cases

### 5. Documentation
- Complete documentation explaining composite unique constraint functionality
- Examples for the reviews table scenario from the GitHub issue
- API behavior documentation, error messages, and supported databases
- Migration guide and troubleshooting section

## Test Results

✅ All new tests pass (11/11)
✅ Composite unique constraint detection works correctly
✅ Validation prevents duplicate combinations while allowing valid ones
✅ Update operations work correctly with constraint checking
✅ Error messages are clear and descriptive

```bash
$ npx jest src/datasources/composite-unique-constraints.test.spec.ts src/datasources/issue-202-integration.test.spec.ts

Test Suites: 2 passed, 2 total
Tests:       11 passed, 11 total
Snapshots:   0 total
Time:        2.148 s
```

## Example Usage

For a reviews table with composite unique constraint `UNIQUE(user_id, server_id)`:

```javascript
// ✅ First review - SUCCESS
POST /api/reviews
{
  "user_id": "user123",
  "server_id": "server456", 
  "rating": 5
}

// ❌ Duplicate review - FAILS with clear error
POST /api/reviews  
{
  "user_id": "user123",
  "server_id": "server456",
  "rating": 4
}
// Error: "Record with user_id, server_id user123, server456 already exists"

// ✅ Same user, different server - SUCCESS
POST /api/reviews
{
  "user_id": "user123", 
  "server_id": "server789",
  "rating": 4
}

// ✅ Different user, same server - SUCCESS  
POST /api/reviews
{
  "user_id": "user456",
  "server_id": "server456", 
  "rating": 3
}
```

## Backward Compatibility

✅ **Fully backward compatible** - existing individual unique column validation continues to work exactly as before
✅ **No breaking changes** - all existing APIs and functionality remain unchanged
✅ **Additive enhancement** - only adds new composite constraint support

## Database Support

| Database | Schema Detection | Validation |
|----------|------------------|------------|
| PostgreSQL | ✅ Full Support | ✅ Full Support |
| MySQL | ✅ Full Support | ✅ Full Support |
| SQL Server | ✅ Full Support | ✅ Full Support |
| MongoDB | ❌ N/A (NoSQL) | ❌ N/A (NoSQL) |
| Airtable | ❌ Not Supported | ❌ Not Supported |

## Files Changed

- `src/types/datasource.types.ts` - Added composite unique constraint types
- `src/datasources/postgres.datasource.ts` - PostgreSQL implementation
- `src/datasources/mysql.datasource.ts` - MySQL implementation  
- `src/datasources/mssql.datasource.ts` - SQL Server implementation
- `src/datasources/mongo.datasource.ts` - MongoDB consistency update
- `src/datasources/airtable.datasource.ts` - Airtable consistency update
- `src/datasources/composite-unique-constraints.test.spec.ts` - Unit tests
- `src/datasources/issue-202-integration.test.spec.ts` - Integration tests
- `docs/composite-unique-constraints.md` - Complete documentation

## Closes

Closes #202

---

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **Documentation**
  * Added comprehensive documentation explaining composite unique constraints, including practical examples, API scenarios, migration guidance, and troubleshooting tips.

* **New Features**
  * Introduced support for composite unique constraints in PostgreSQL, MySQL, and SQL Server datasources, enabling enforcement of uniqueness across multiple columns.
  * Enhanced error messages to specify which column combinations caused uniqueness violations.
  * Included composite unique constraints property in schema for Airtable and MongoDB with clarifications on support limitations.

* **Bug Fixes**
  * Ensured composite unique constraints are properly detected and validated during record creation and updates.

* **Tests**
  * Added new unit and integration tests to verify correct enforcement of composite unique constraints across supported databases.

* **Style**
  * Applied formatting and whitespace improvements across middleware, services, and bootup code for better readability.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

vishwamartur committed:
feat: Add DataSourceCompositeUniqueConstraint interface for #202

- Add DataSourceCompositeUniqueConstraint interface to support multi-column unique constraints
- Extend DataSourceSchema to include composite_unique_constraints field
- This enables detection and validation of composite unique constraints across all datasources

vishwamartur committed:
feat: Implement composite unique constraints for PostgreSQL datasource (#202)

- Add schema discovery for composite unique constraints using information_schema
- Enhance uniqueCheck method to validate composite constraints
- Remove early return when no individual unique columns exist
- Build proper WHERE clauses for multi-column uniqueness validation
- Support updates by excluding current record ID from constraint checks

vishwamartur committed:
feat: Implement composite unique constraints for MySQL datasource (#202)

- Add schema discovery for composite unique constraints using INFORMATION_SCHEMA
- Enhance uniqueCheck method to validate composite constraints with proper logging
- Build WHERE clauses for multi-column uniqueness validation
- Support updates by excluding current record ID from constraint checks

vishwamartur committed:
feat: Implement composite unique constraints for MSSQL datasource (#202)

- Add schema discovery for composite unique constraints using INFORMATION_SCHEMA
- Enhance uniqueCheck method to validate composite constraints
- Remove early return when no individual unique columns exist
- Build WHERE clauses for multi-column uniqueness validation with proper SQL Server syntax
- Support updates by excluding current record ID from constraint checks

vishwamartur committed:
feat: Add composite unique constraints support for MongoDB and Airtable (#202)

- Add empty composite_unique_constraints array to MongoDB datasource schema
- Add empty composite_unique_constraints array to Airtable datasource schema
- MongoDB and Airtable don't support SQL-style composite unique constraints
- This maintains schema consistency across all datasource types

vishwamartur committed:
test: Add comprehensive test suite for composite unique constraints (#202)

- Add unit tests for PostgreSQL, MySQL, and MSSQL composite unique constraint validation
- Add integration tests that specifically test the GitHub issue #202 scenario
- Test reviews table with user_id and server_id composite constraint
- Verify constraint violation detection, updates, and edge cases
- All tests pass and demonstrate the fix works correctly

vishwamartur committed:
docs: Add comprehensive documentation for composite unique constraints (#202)

- Add detailed documentation explaining composite unique constraint functionality
- Include examples for reviews table scenario from GitHub issue #202
- Document API behavior, error messages, and supported databases
- Provide migration guide and troubleshooting section
- Include technical implementation details and schema structure

andyslack commented:
@vishwamartur this looks fantastic,

I am getting some `lint` errors:

```
/Users/andrewslack/Sites/JuicyLlama/llana/src/datasources/airtable.datasource.ts
  16:2  error  'DataSourceCompositeUniqueConstraint' is defined but never used  @typescript-eslint/no-unused-vars

/Users/andrewslack/Sites/JuicyLlama/llana/src/datasources/mongo.datasource.ts
  17:2  error  'DataSourceCompositeUniqueConstraint' is defined but never used  @typescript-eslint/no-unused-vars

/Users/andrewslack/Sites/JuicyLlama/llana/src/datasources/mssql.datasource.ts
  17:2  error  'DataSourceCompositeUniqueConstraint' is defined but never used  @typescript-eslint/no-unused-vars

/Users/andrewslack/Sites/JuicyLlama/llana/src/datasources/mysql.datasource.ts
  17:2  error  'DataSourceCompositeUniqueConstraint' is defined but never used  @typescript-eslint/no-unused-vars

/Users/andrewslack/Sites/JuicyLlama/llana/src/datasources/postgres.datasource.ts
  16:2  error  'DataSourceCompositeUniqueConstraint' is defined but never used  @typescript-eslint/no-unused-vars
  ```
  
  Can you either remove the imports or use them in the files?

vishwamartur committed:
fix: Resolve lint errors for unused DataSourceCompositeUniqueConstraint imports

- Remove unused imports from airtable.datasource.ts and mongo.datasource.ts
- Add explicit type annotations to postgres.datasource.ts, mysql.datasource.ts, and mssql.datasource.ts
- This makes the DataSourceCompositeUniqueConstraint type usage explicit and resolves ESLint no-unused-vars errors
- All tests still pass (11/11) and build completes successfully

Addresses feedback from @andyslack in PR #269

vishwamartur commented:
<img width="2156" height="838" alt="image" src="https://github.com/user-attachments/assets/36bbbcc1-a297-4432-8757-e080cabe7d06" />
